src/main/java/uk/ac/cam/cl/dtg/segue/dao/users/PgUsers.java

```diff
@@ -493,23 +493,26 @@ public Map<Gender, Long> getGenderCount() throws SegueDatabaseException {
     }
 
     @Override
-    public Map<Role, Long> getRolesLastSeenOver(final TimeInterval timeInterval) throws SegueDatabaseException {
-        String query = "SELECT role, count(1) FROM users WHERE NOT deleted AND last_seen >= now() - ? GROUP BY role";
-        try (Connection conn = database.getDatabaseConnection();
-             PreparedStatement pst = conn.prepareStatement(query)
-        ) {
-            pst.setObject(FIELD_GET_PERIOD_ROLES_INTERVAL, timeInterval.getPGInterval());
-
-            try (ResultSet results = pst.executeQuery()) {
-                Map<Role, Long> resultsToReturn = Maps.newHashMap();
-                while (results.next()) {
-                    resultsToReturn.put(Role.valueOf(results.getString("role")), results.getLong("count"));
+    public Map<String, Map<Role, Long>> getRolesLastSeenOver(final String[] timeRanges) throws SegueDatabaseException {
```

Comment:
If I understood this right, it looks like what was causing the code to be slow was having to open and close connections multiple times, which is why this function now accepts an array, rather than a single value. So that you can share one connection. This is fine.

Then you're also returning the results grouped by a string, which is in the same format as is expected by StatisticsManager (sevenDays, sixMonths, etc). This is not right, and this needs changing. This function needs to group things by `TimeInterval`. Then, in `StatisticsManager` you make the conversion from `TimeInterval` to the right string. The way you've done it is less robust, and this function does not follow the single responsibility principle (this function is doing too many things).

Reply:
Changes have been made to follow single responsibility principle.